hail/hail/ir-gen/src/Main.scala

```diff
@@ -0,0 +1,1108 @@
+    def withTraits(newTraits: Trait*): IR = copy(traits = traits ++ newTraits)
+    def withMethod(methodDef: String): IR = copy(extraMethods = extraMethods :+ methodDef)
+    def typed(typ: String): IR = withTraits(TypedIR(typ))
+    def withConstraint(c: String): IR = copy(constraints = constraints :+ c)
+
+    def withCompanionExtension: IR = copy(hasCompanionExtension = true)
+
+    def withClassExtension: IR = withTraits(Trait(s"${name}Ext"))
+
+    def withDocstring(docstring: String): IR = copy(docstring = docstring)
+
+    private def nChildren: NChildren = attsAndChildren.foldLeft(NChildren())(_ + _.nChildren)
+
+    private def children: String =
+      attsAndChildren.foldLeft[ChildrenSeq](ChildrenSeq.empty)(_ ++ _.childrenSeq).toString
+
+    private def childrenOffsets: Seq[NChildren] =
+      attsAndChildren.scanLeft(NChildren())(_ + _.nChildren)
+
+    private def copyMethod: String = {
+      val decl = s"override def copyWithNewChildren(newChildren: IndexedSeq[BaseIR]): $name = "
+      val assertion = s"assert(newChildren.length == $nChildren)"
+      val body = name + attsAndChildren.zipWithIndex.map { case (x, i) =>
+        try
+          x.pack.copyWithNewChildren(
+            x.name,
+            ChildrenSeqSlice(
+              ChildrenSeq.Dynamic("newChildren"),
+              nChildren,
+              childrenOffsets(i),
+              x.nChildren,
+            ),
+          )
+        catch {
+          case _: NotImplementedError =>
+            assert(false, name)
+        }
+      }.mkString("(", ", ", ")")
+      decl + "{\n    " + assertion + "\n    " + body + "}"
+    }
+
+    private def paramList = s"$name(${attsAndChildren.map(_.generateDeclaration).mkString(", ")})"
+
+    private def classDecl =
+      s"final case class $paramList extends IR" + traits.map(" with " + _.name).mkString
+
+    private def classBody = {
+      val childrenSeqDef = if (nChildren.hasStaticValue(0))
+        s"override def childrenSeq: IndexedSeq[BaseIR] = IndexedSeq.empty"
+      else
+        s"override lazy val childrenSeq: IndexedSeq[BaseIR] = $children"
+      val extraMethods =
+        this.extraMethods :+ childrenSeqDef :+ copyMethod
+      val constraints = this.constraints ++ attsAndChildren.flatMap(_.constraints)
+      if (constraints.nonEmpty || extraMethods.nonEmpty) {
+        (
+          " {" +
+            (if (constraints.nonEmpty)
+               constraints.map(c => s"  require($c)").mkString("\n", "\n", "\n")
+             else "")
+            + (
+              if (extraMethods.nonEmpty)
+                extraMethods.map("  " + _).mkString("\n", "\n", "\n")
+              else ""
+            )
+            + "}"
+        )
+      } else ""
+    }
+
+    private def classDef =
+      (if (docstring.nonEmpty) s"\n/** $docstring*/\n" else "") + classDecl + classBody
+
+    private def companionDef = if (hasCompanionExtension)
+      s"object $name extends ${name}CompanionExt\n"
+    else ""
```

Comment:
maybe make these locals in `generateDef`

Reply:
I tried it, but if I move `classDef` inside, then it feels like `classBody`, and by extension `copyMethod` and others, should also be inside. But then `generateDef` is too big and deeply nested.